sunpy/map/mapbase.py

```diff
@@ -2584,6 +2587,74 @@ def _check_axes(self, axes, warn_different_wcs=False):
 
         return axes
 
+    def reproject_to(self, target_wcs, algorithm='interpolation', **reproject_args):
+        """
+        Reproject the map to a different WCS
+
+        .. note::
+            This method requires the optional package `reproject` to be installed.
+
+        Parameters
+        ----------
+        target_wcs : `dict` or `~astropy.wcs.WCS`
+            The destination FITS WCS header or WCS instance
+        algorithm : str
+            One of the supported `reproject` algorithms (see below)
+
+        Returns
+        -------
+        outmap : `~sunpy.map.GenericMap`
+            The reprojected map
+
+        Notes
+        -----
+        The reprojected map does not preserve any metadata beyond the WCS-associated
+        metadata.
+
+        The supported `reproject` algorithms are:
+
+        * 'interpolation' for :func:`~reproject.reproject_interp`
+        * 'adaptive' for :func:`~reproject.reproject_adaptive`
+        * 'exact' for :func:`~reproject.reproject_exact`
+
+        See the respective documentation for these functions for additional keyword
+        arguments that are allowed.
+
+        .. minigallery:: sunpy.map.GenericMap.reproject_to
+        """
+        try:
+            import reproject
+        except ImportError:
+            raise ImportError("This method requires the optional package `reproject`.")
+
+        if not isinstance(target_wcs, astropy.wcs.WCS):
+            target_wcs = astropy.wcs.WCS(target_wcs)
```

Comment:
This is not the best way around. It would be better to check if it's a header-like object (`isinstance(target_wcs, collections.Mapping)`).

https://github.com/sunpy/ndcube/blob/2dff8c1d90f1c43b1eafc3be4c0bd86ad04b8957/ndcube/ndcube.py#L783-L784

Reply:
Why is that better?  I need a `WCS` instance for the rest of this code.  I don't care the type of the input as long as the `WCS()` constructor works.

Reply:
Because reproject supports WCSes which aren't of type `astropy.wcs.WCS`. Sure *for now* map doesn't support that but why not write it the way around which is more correct and future proof?

Reply:
Given we will be adopting the ndcube version before we support any other type of WCS this is a real irrelevant nit.

Reply:
Ah, I see what you mean.  I was approaching it from the perspective that I needed it to be `astropy.wcs.WCS` to rely on the existence of `.array_shape` (and, as currently written, `.to_header()`), and that maybe the constructor would accept other inputs than `Mapping` in the future.  Sounds like we can deal with the future possibilities in the future.

Reply:
`.array_shape` is an APE 14 property, so that one is fine. `to_header` is `wcs.WCS` specific obviously.